File: mobile/modules/bluetooth-sdk/android/src/main/java/com/mentra/bluetoothsdk/drivers/DeviceHostImpl.kt
package com.mentra.bluetoothsdk.drivers

import com.mentra.bluetoothsdk.BluetoothSdkDefaults
import com.mentra.bluetoothsdk.Bridge
import com.mentra.bluetoothsdk.DeviceManager
import com.mentra.bluetoothsdk.DeviceStore
import com.mentra.bluetoothsdk.utils.ConnTypes

/**
 * Concrete [DeviceHost] that routes a driver's callbacks into the real app
 * singletons (Bridge / DeviceStore / DeviceManager) — i.e. exactly the inline
 * calls RemoteHarness used to make, moved behind the façade. See the mapping
 * table in docs/device-driver-contract.md §6.
 *
 * @param deviceType the driver's type string, used as the model on touch events
 *                   (matches what RemoteHarness passed before).
 */
class DeviceHostImpl(private val deviceType: String) : DeviceHost {

    override fun reset() {
        DeviceStore.apply("glasses", "fullyBooted", false)
        DeviceStore.apply("glasses", "connected", false)
        DeviceStore.apply("glasses", "connectionState", ConnTypes.CONNECTING)
        DeviceStore.apply("glasses", "micEnabled", false)
        DeviceStore.apply(
            "glasses",
            "voiceActivityDetectionEnabled",
            BluetoothSdkDefaults.VOICE_ACTIVITY_DETECTION_ENABLED,
        )
    }

    override fun reportConnectionState(state: ConnectionState) {
        when (state) {
            ConnectionState.CONNECTED -> {
                DeviceStore.apply("glasses", "connected", true)
                DeviceStore.apply("glasses", "connectionState", ConnTypes.CONNECTED)
            }
            ConnectionState.CONNECTING -> {
                DeviceStore.apply("glasses", "connected", false)
                DeviceStore.apply("glasses", "connectionState", ConnTypes.CONNECTING)
            }
            ConnectionState.DISCONNECTED -> {
                DeviceStore.apply("glasses", "connected", false)
                DeviceStore.apply("glasses", "connectionState", ConnTypes.DISCONNECTED)
            }
        }
    }

    override fun reportReady(ready: Boolean) {
        DeviceStore.apply("glasses", "fullyBooted", ready)
    }

    override fun reportDeviceInfo(info: DeviceInfo) {
        DeviceStore.apply("glasses", "deviceModel", info.model)
        info.serial?.let { DeviceStore.apply("glasses", "serialNumber", it) }
        info.firmware?.let { DeviceStore.apply("glasses", "firmwareVersion", it) }
        info.color?.let { DeviceStore.apply("glasses", "color", it) }
        info.style?.let { DeviceStore.apply("glasses", "style", it) }
    }

    override fun emitBattery(level: Int, charging: Boolean) {
        DeviceStore.apply("glasses", "batteryLevel", level)
        DeviceStore.apply("glasses", "charging", charging)
        Bridge.sendBatteryStatus(level, charging)
    }

    override fun emitMicAudio(lc3: ByteArray, frameSize: Int) {
        DeviceManager.getInstance().handleGlassesMicData(lc3, frameSize)
    }

    override fun reportMicEnabled(on: Boolean) {
        DeviceStore.apply("glasses", "micEnabled", on)
    }

    override fun emitTouchEvent(gesture: String) {
        Bridge.sendTouchEvent(deviceType, gesture, System.currentTimeMillis(), 0)
    }

    override fun emitImu(
        accel: DoubleArray,
        gyro: DoubleArray,
        mag: DoubleArray,
        quat: DoubleArray,
        euler: DoubleArray,
    ) {
        Bridge.sendImuDataEvent(accel, gyro, mag, quat, euler, System.currentTimeMillis())
    }

    override fun reportCommandResult(requestId: String, ok: Boolean, error: String?) {
        Bridge.sendRgbLedControlResponse(requestId, ok, error)
    }

    override fun reportDiscoveredDevice(id: String, name: String) {
        Bridge.sendDiscoveredDevice(id, name)
    }

    override fun log(msg: String) {
        Bridge.log(msg)
    }
}

File: mobile/modules/bluetooth-sdk/ios/Source/drivers/DeviceHostImpl.swift
//
//  DeviceHostImpl.swift
//  Concrete DeviceHost (iOS) — routes a driver's callbacks into the real app
//  singletons (Bridge / DeviceStore / DeviceManager), i.e. exactly the inline
//  calls RemoteHarness.swift used to make, moved behind the façade. Mirror of
//  the Android DeviceHostImpl.kt. See docs/device-driver-contract.md §6.
//

import Foundation

@MainActor
final class DeviceHostImpl: DeviceHost {
    /// The driver's type string, used as the model on touch events (matches what
    /// RemoteHarness passed before).
    private let deviceType: String

    init(deviceType: String) {
        self.deviceType = deviceType
    }

    func reset() {
        DeviceStore.shared.apply("glasses", "fullyBooted", false)
        DeviceStore.shared.apply("glasses", "connected", false)
        DeviceStore.shared.apply("glasses", "connectionState", ConnTypes.CONNECTING)
        DeviceStore.shared.apply("glasses", "micEnabled", false)
        DeviceStore.shared.apply(
            "glasses",
            "voiceActivityDetectionEnabled",
            BluetoothSdkDefaults.voiceActivityDetectionEnabled
        )
    }

    func reportConnectionState(_ state: ConnectionState) {
        switch state {
        case .connected:
            DeviceStore.shared.apply("glasses", "connected", true)
            DeviceStore.shared.apply("glasses", "connectionState", ConnTypes.CONNECTED)
        case .connecting:
            DeviceStore.shared.apply("glasses", "connected", false)
            DeviceStore.shared.apply("glasses", "connectionState", ConnTypes.CONNECTING)
        case .disconnected:
            DeviceStore.shared.apply("glasses", "connected", false)
            DeviceStore.shared.apply("glasses", "connectionState", ConnTypes.DISCONNECTED)
        }
    }

    func reportReady(_ ready: Bool) {
        DeviceStore.shared.apply("glasses", "fullyBooted", ready)
    }

    func reportDeviceInfo(_ info: DeviceInfo) {
        DeviceStore.shared.apply("glasses", "deviceModel", info.model)
        if let serial = info.serial { DeviceStore.shared.apply("glasses", "serialNumber", serial) }
        if let firmware = info.firmware { DeviceStore.shared.apply("glasses", "firmwareVersion", firmware) }
        if let color = info.color { DeviceStore.shared.apply("glasses", "color", color) }
        if let style = info.style { DeviceStore.shared.apply("glasses", "style", style) }
    }

    func emitBattery(level: Int, charging: Bool) {
        DeviceStore.shared.apply("glasses", "batteryLevel", level)
        DeviceStore.shared.apply("glasses", "charging", charging)
        Bridge.sendBatteryStatus(level: level, charging: charging)
    }

    func emitMicAudio(_ lc3: Data, frameSize: Int) {
        DeviceManager.shared.handleGlassesMicData(lc3, frameSize)
    }

    func reportMicEnabled(_ on: Bool) {
        DeviceStore.shared.apply("glasses", "micEnabled", on)
    }

    func emitTouchEvent(_ gesture: String) {
        Bridge.sendTouchEvent(deviceModel: deviceType, gestureName: gesture, timestamp: Int64(Date().timeIntervalSince1970 * 1000))
    }

    func emitImu(accel: [Double]) {
        let x = accel.count > 0 ? accel[0] : 0
        let y = accel.count > 1 ? accel[1] : 0
        let z = accel.count > 2 ? accel[2] : 0
        Bridge.sendAccelEvent(x: Float(x), y: Float(y), z: Float(z), timestamp: Int64(Date().timeIntervalSince1970 * 1000))
    }

    func reportCommandResult(requestId: String, ok: Bool, error: String?) {
        Bridge.sendRgbLedControlResponse(requestId: requestId, success: ok, error: error)
    }

    func reportDiscoveredDevice(id: String, name: String) {
        Bridge.sendDiscoveredDevice(id, name)
    }

    func log(_ msg: String) {
        Bridge.log(msg)
    }
}
